iohub/dataset_base.py

```diff
@@ -0,0 +1,71 @@
+from abc import ABC, abstractmethod
+from typing import Generator, Union
+
+import numpy as np
+from numpy.typing import ArrayLike
+
+
+class BaseFOV(ABC):
+    # NOTE: not using the `data` method from ngff Position
+
+    @property
+    @abstractmethod
+    def channels(self) -> list[str]:
+        raise NotImplementedError
+
+    def channel_index(self, key: str) -> int:
+        """Return index of given channel."""
+        return self.channels.index(key)
+
+    @property
+    @abstractmethod
+    def shape(self) -> tuple[int, int, int, int, int]:
+        # NOTE: suggestion, fix dimension to 5?
+        # We could me more restrictive than ome-zarr
+        raise NotImplementedError
+
+    @abstractmethod
+    def __getitem__(
+        self,
+        key: Union[int, slice, tuple[Union[int, slice], ...]],
+    ) -> ArrayLike:
+        """
+        Output object must support __array__ interface, np.asarray(...).
+        """
+        raise NotImplementedError
+
+    @property
+    @abstractmethod
+    def ndim(self) -> int:
+        raise NotImplementedError
+
+    @property
+    @abstractmethod
+    def dtype(self) -> np.dtype:
```

Comment:
You can get `DTypeLike` from `numpy.typing`. So we don't need the np namespace in this file.

Reply:
`DTypeLike` is a bit broader than `np.dtype`, for example, the string `"unit8"` is `DTypeLike`, if that's OK with you I'm fine with switching.

Reply:
IMO it's fine since that seems to be how most array libraries (numpy, zarr, torch etc) type their API. But stricter typing doesn't harm so I don't have a strong opinion here.